File: AgencyCampaign/AgencyCampaign.Domain/Entities/CreatorPayment.cs
using AgencyCampaign.Domain.ValueObjects;
using Archon.Core.Entities;

namespace AgencyCampaign.Domain.Entities
{
    public sealed class CreatorPayment : Entity
    {
        private readonly List<CreatorPaymentEvent> events = [];

        public long CampaignCreatorId { get; private set; }

        public CampaignCreator? CampaignCreator { get; private set; }

        public long CreatorId { get; private set; }

        public Creator? Creator { get; private set; }

        public long? CampaignDocumentId { get; private set; }

        public CampaignDocument? CampaignDocument { get; private set; }

        public decimal GrossAmount { get; private set; }

        public decimal Discounts { get; private set; }

        // Imposto retido na fonte ao pagar o creator (IRRF/INSS/ISS conforme regime). Por ora informado
        // pelo operador (calculo automatico = fase 2); entra no liquido e no relatorio de retencoes.
        public decimal TaxWithheld { get; private set; }

        public decimal NetAmount { get; private set; }

        public string? Description { get; private set; }

        public PaymentMethod Method { get; private set; } = PaymentMethod.Pix;

        public PaymentStatus Status { get; private set; } = PaymentStatus.Pending;

        public string? Provider { get; private set; }

        public string? ProviderTransactionId { get; private set; }

        public string? EndToEndId { get; private set; }

        public string? IdempotencyKey { get; private set; }

        public string? PixKey { get; private set; }

        public PixKeyType? PixKeyType { get; private set; }

        public string? InvoiceNumber { get; private set; }

        public string? InvoiceUrl { get; private set; }

        public DateTimeOffset? InvoiceIssuedAt { get; private set; }

        public DateTimeOffset? ScheduledFor { get; private set; }

        public DateTimeOffset? PaidAt { get; private set; }

        public DateTimeOffset? FailedAt { get; private set; }

        public string? FailureReason { get; private set; }

        // Maker-checker: quem criou e quem aprovou o pagamento. O aprovador precisa ser diferente do criador
        // (segregacao de funcoes); acima do teto da agencia, o repasse so e agendado apos a aprovacao.
        public long? CreatedByUserId { get; private set; }

        public DateTimeOffset? ApprovedAt { get; private set; }

        public long? ApprovedByUserId { get; private set; }

        public bool IsApproved => ApprovedAt.HasValue;

        public IReadOnlyCollection<CreatorPaymentEvent> Events => events.AsReadOnly();

        private CreatorPayment()
        {
        }

        public CreatorPayment(long campaignCreatorId, long creatorId, decimal grossAmount, decimal discounts, PaymentMethod method, string? description = null, long? campaignDocumentId = null, decimal taxWithheld = 0)
        {
            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(campaignCreatorId);
            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(creatorId);
            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(grossAmount);
            ArgumentOutOfRangeException.ThrowIfNegative(discounts);
            ArgumentOutOfRangeException.ThrowIfNegative(taxWithheld);

            if (discounts + taxWithheld > grossAmount)
            {
                throw new ArgumentException("Discounts and tax withheld cannot exceed gross amount.", nameof(discounts));
            }

            CampaignCreatorId = campaignCreatorId;
            CreatorId = creatorId;
            CampaignDocumentId = campaignDocumentId;
            GrossAmount = grossAmount;
            Discounts = discounts;
            TaxWithheld = taxWithheld;
            NetAmount = grossAmount - discounts - taxWithheld;
            Method = method;
            Description = Normalize(description);
        }

        public void Update(decimal grossAmount, decimal discounts, PaymentMethod method, string? description, decimal taxWithheld = 0)
        {
            EnsureMutable();

            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(grossAmount);
            ArgumentOutOfRangeException.ThrowIfNegative(discounts);
            ArgumentOutOfRangeException.ThrowIfNegative(taxWithheld);

            if (discounts + taxWithheld > grossAmount)
            {
                throw new ArgumentException("Discounts and tax withheld cannot exceed gross amount.", nameof(discounts));
            }

            GrossAmount = grossAmount;
            Discounts = discounts;
            TaxWithheld = taxWithheld;
            NetAmount = grossAmount - discounts - taxWithheld;
            Method = method;
            Description = Normalize(description);
        }

        public void SnapshotPixDestination(string pixKey, PixKeyType pixKeyType)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(pixKey);
            PixKey = pixKey.Trim();
            PixKeyType = pixKeyType;
        }

        public void AttachInvoice(string? invoiceNumber, string? invoiceUrl, DateTimeOffset? issuedAt)
        {
            InvoiceNumber = Normalize(invoiceNumber);
            InvoiceUrl = Normalize(invoiceUrl);
            InvoiceIssuedAt = issuedAt?.ToUniversalTime();
        }

        public void AttachToProvider(string provider, string providerTransactionId)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(provider);
            ArgumentException.ThrowIfNullOrWhiteSpace(providerTransactionId);

            Provider = provider.Trim();
            ProviderTransactionId = providerTransactionId.Trim();
        }

        // EndToEndId (e2eId) do Pix: identificador imutavel atribuido pelo Banco Central, usado para
        // conciliar o repasse contra o extrato e comprovar o pagamento ao creator.
        public void AttachEndToEndId(string endToEndId)
        {
            EndToEndId = Normalize(endToEndId);
        }

        // Chave de idempotencia do payout: gerada no agendamento e enviada ao IntegrationPlatform para
        // que um retry de rede/reprocessamento nao dispare o Pix duas vezes.
        public void AssignIdempotencyKey(string idempotencyKey)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(idempotencyKey);
            IdempotencyKey = idempotencyKey.Trim();
        }

        public void Schedule(DateTimeOffset scheduledFor)
        {
            EnsureMutable();

            ScheduledFor = scheduledFor.ToUniversalTime();
            Status = PaymentStatus.Scheduled;
        }

        public void SetCreatedBy(long? userId)
        {
            CreatedByUserId = userId;
        }

        // Aprovacao (maker-checker): o aprovador deve ser diferente de quem criou; nao reaprova nem aprova
        // pagamento ja finalizado.
        public void Approve(long approverUserId)
        {
            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(approverUserId);

            if (Status == PaymentStatus.Paid || Status == PaymentStatus.Cancelled)
            {
                throw new InvalidOperationException("creatorPayment.alreadyFinalized");
            }

            if (ApprovedAt.HasValue)
            {
                throw new InvalidOperationException("creatorPayment.alreadyApproved");
            }

            if (CreatedByUserId.HasValue && CreatedByUserId.Value == approverUserId)
            {
                throw new InvalidOperationException("creatorPayment.approverMustDiffer");
            }

            ApprovedAt = DateTimeOffset.UtcNow;
            ApprovedByUserId = approverUserId;
        }

        public void MarkPaid(DateTimeOffset paidAt)
        {
            if (Status == PaymentStatus.Cancelled)
            {
                throw new InvalidOperationException("creatorPayment.cannotPayCancelled");
            }

            PaidAt = paidAt.ToUniversalTime();
            Status = PaymentStatus.Paid;
            FailureReason = null;
            FailedAt = null;
        }

        public void MarkFailed(string? reason, DateTimeOffset? failedAt = null)
        {
            if (Status == PaymentStatus.Paid || Status == PaymentStatus.Cancelled)
            {
                throw new InvalidOperationException("creatorPayment.cannotFailFinalized");
            }

            FailureReason = Normalize(reason);
            FailedAt = (failedAt ?? DateTimeOffset.UtcNow).ToUniversalTime();
            Status = PaymentStatus.Failed;
        }

        public void Cancel(string? reason = null)
        {
            EnsureNotFinalized();
            FailureReason = Normalize(reason);
            Status = PaymentStatus.Cancelled;
        }

        public CreatorPaymentEvent RegisterEvent(CreatorPaymentEventType eventType, string? description = null, string? metadata = null, DateTimeOffset? occurredAt = null)
        {
            CreatorPaymentEvent evt = new(Id, eventType, description, metadata, occurredAt);
            events.Add(evt);
            return evt;
        }

        private void EnsureMutable()
        {
            if (Status == PaymentStatus.Paid || Status == PaymentStatus.Cancelled)
            {
                throw new InvalidOperationException("creatorPayment.alreadyFinalized");
            }
        }

        private void EnsureNotFinalized()
        {
            if (Status == PaymentStatus.Paid)
            {
                throw new InvalidOperationException("creatorPayment.cannotCancelPaid");
            }
        }

        private static string? Normalize(string? value)
        {
            return string.IsNullOrWhiteSpace(value) ? null : value.Trim();
        }
    }
}

File: AgencyCampaign/AgencyCampaign.Web/src/types/creatorPayment.ts
export const PaymentStatus = {
  Pending: 1,
  Scheduled: 2,
  Paid: 3,
  Failed: 4,
  Cancelled: 5,
} as const
export type PaymentStatusValue = (typeof PaymentStatus)[keyof typeof PaymentStatus]

export const paymentStatusLabels: Record<PaymentStatusValue, string> = {
  1: 'Pendente',
  2: 'Agendado',
  3: 'Pago',
  4: 'Falhou',
  5: 'Cancelado',
}

export const PaymentMethod = {
  Pix: 1,
  Ted: 2,
  Manual: 3,
} as const
export type PaymentMethodValue = (typeof PaymentMethod)[keyof typeof PaymentMethod]

export const paymentMethodLabels: Record<PaymentMethodValue, string> = {
  1: 'PIX',
  2: 'TED',
  3: 'Manual',
}

export const PixKeyType = {
  Cpf: 1,
  Cnpj: 2,
  Email: 3,
  Phone: 4,
  Random: 5,
} as const
export type PixKeyTypeValue = (typeof PixKeyType)[keyof typeof PixKeyType]

export const pixKeyTypeLabels: Record<PixKeyTypeValue, string> = {
  1: 'CPF',
  2: 'CNPJ',
  3: 'E-mail',
  4: 'Telefone',
  5: 'Aleatória',
}

export const CreatorPaymentEventType = {
  Created: 1,
  Updated: 2,
  Scheduled: 3,
  ProviderAccepted: 4,
  Paid: 5,
  Failed: 6,
  Cancelled: 7,
  InvoiceAttached: 8,
  ProviderSyncError: 9,
} as const
export type CreatorPaymentEventTypeValue =
  (typeof CreatorPaymentEventType)[keyof typeof CreatorPaymentEventType]

export const creatorPaymentEventTypeLabels: Record<CreatorPaymentEventTypeValue, string> = {
  1: 'Criado',
  2: 'Atualizado',
  3: 'Agendado',
  4: 'Aceito pelo provider',
  5: 'Pago',
  6: 'Falhou',
  7: 'Cancelado',
  8: 'NF anexada',
  9: 'Erro de sincronização',
}

export interface CreatorPaymentEvent {
  id: number
  eventType: CreatorPaymentEventTypeValue
  occurredAt: string
  description?: string
  metadata?: string
}

export interface CreatorPayment {
  id: number
  campaignCreatorId: number
  creatorId: number
  creatorName?: string
  creatorPixKey?: string
  creatorPixKeyType?: PixKeyTypeValue
  campaignId?: number
  campaignName?: string
  campaignDocumentId?: number
  grossAmount: number
  discounts: number
  netAmount: number
  description?: string
  method: PaymentMethodValue
  status: PaymentStatusValue
  provider?: string
  providerTransactionId?: string
  pixKey?: string
  pixKeyType?: PixKeyTypeValue
  invoiceNumber?: string
  invoiceUrl?: string
  invoiceIssuedAt?: string
  scheduledFor?: string
  paidAt?: string
  failedAt?: string
  failureReason?: string
  createdAt: string
  updatedAt?: string
  events: CreatorPaymentEvent[]
}
